kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/punctuators/AbstractThrottledPunctuator.java

```diff
@@ -0,0 +1,133 @@
+package org.hypertrace.core.kafkastreams.framework.punctuators;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.hypertrace.core.kafkastreams.framework.punctuators.action.TaskResult;
+
+@Slf4j
+public abstract class AbstractThrottledPunctuator<T> implements Punctuator {
+  private final Clock clock;
+  private final KeyValueStore<Long, ArrayList<T>> objectStore;
+  private final ThrottledPunctuatorConfig config;
+
+  public AbstractThrottledPunctuator(
+      Clock clock,
+      ThrottledPunctuatorConfig config,
+      KeyValueStore<Long, ArrayList<T>> objectStore) {
+    this.clock = clock;
+    this.config = config;
+    this.objectStore = objectStore;
+  }
+
+  public void scheduleTask(long atTimestampInMs, T object) {
```

Comment:
Suggestion :
Since we have exposed these methods which is not being used anywhere in the current punctuator implementation, it is no longer just a punctuator. It is wrapper for state store having windowed timestamp as keys, as well as punctuation that run on top of that state store.

Ideally we should pull out operation on object store to a separate class, and keep it as a simple Punctuator implementation.